#### Import Libraries 

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

#### Reading Dataset

In [2]:
hotels = pd.read_csv('zomato.csv')
hotels.head()

address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate votes                             phone  \
0          Yes        Yes  4.1/5   775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5   787                      080 41714161   
2          Yes         No  3.8/5   918                    +91 9663487993   
3           No         No  3.7/5    88                    +91 9620009302   
4           No         No  3.8/5   166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet  
1     Chinese, North Indian, Thai                         800          Buffet  
2          Cafe, Mexican, Italian                         800          Buffet  
3      South Indian, North Indian                         300          Buffet  
4        North Indian, Rajasthani                         600          Buffet

#### General insights about the data 

In [3]:
# Display the number of rows and columns

hotels.shape

(56252, 13)

- There are 56252 rows and 13 columns in the dataset.

In [4]:
# Dataset Info

hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56252 entries, 0 to 56251
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   address                      56235 non-null  object
 1   name                         56236 non-null  object
 2   online_order                 56233 non-null  object
 3   book_table                   56194 non-null  object
 4   rate                         48414 non-null  object
 5   votes                        56174 non-null  object
 6   phone                        54956 non-null  object
 7   location                     56126 non-null  object
 8   rest_type                    55914 non-null  object
 9   dish_liked                   28027 non-null  object
 10  cuisines                     56049 non-null  object
 11  approx_cost(for two people)  55731 non-null  object
 12  listed_in(type)              51642 non-null  object
dtypes: object(13)
memory usage: 5.6

- It shows all the variables are of string/objective data type.
- Out of which Votes and Approx_cost columns are holding integer value. 

In [5]:
# Check missing values

hotels.isnull().sum()

address                           17
name                              16
online_order                      19
book_table                        58
rate                            7838
votes                             78
phone                           1296
location                         126
rest_type                        338
dish_liked                     28225
cuisines                         203
approx_cost(for two people)      521
listed_in(type)                 4610
dtype: int64

In [6]:
# Calculate the percentage of missing values in each column

hotels.isna().mean().round(4) * 100

address                         0.03
name                            0.03
online_order                    0.03
book_table                      0.10
rate                           13.93
votes                           0.14
phone                           2.30
location                        0.22
rest_type                       0.60
dish_liked                     50.18
cuisines                        0.36
approx_cost(for two people)     0.93
listed_in(type)                 8.20
dtype: float64

- The percentage of missing values for the columns Dish_liked and Rate is ~50% and ~14% respectively.

#### Data pre-processing 

In [7]:
# Removing Unwanted Columns

hotels.columns

Index(['address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'listed_in(type)'],
      dtype='object')

In [8]:
hotels.drop(['address','phone'],axis=1,inplace=True)
hotels.head()

name online_order book_table   rate votes      location  \
0                  Jalsa          Yes        Yes  4.1/5   775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5   787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5   918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5    88  Banashankari   
4          Grand Village           No         No  3.8/5   166  Basavanagudi   

             rest_type                                         dish_liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people) listed_in(type)  
0  North Indian, Mughlai, Chinese                         800          Buffet  
1     Chinese, North Indian, Thai                         800          Buffet  
2          Cafe, Mexican, Italian                         800          Buffet  
3      South Indian, North Indian                         300          Buffet  
4        North Indian, Rajasthani                         600          Buffet

In [9]:
# Renaming and Selecting Columns in a Dataset

hotels.rename(columns = {'name':'Name', 'online_order':'Online_Order', 'book_table':'Book_Table', 'rate':'Rating', 'votes':'Votes', 'location':'Location', 'rest_type':'Rest_Type', 'dish_liked':'Dish_Liked', 'cuisines':'Cuisines', 'approx_cost(for two people)':'Approx_Cost', 'listed_in(type)':'Type'},inplace=True)
hotels.head()

Name Online_Order Book_Table Rating Votes      Location  \
0                  Jalsa          Yes        Yes  4.1/5   775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5   787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5   918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5    88  Banashankari   
4          Grand Village           No         No  3.8/5   166  Basavanagudi   

             Rest_Type                                         Dish_Liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   

                         Cuisines Approx_Cost    Type  
0  North Indian, Mughlai, Chinese         800  Buffet  
1     Chinese, North Indian, Thai         800  Buffet  
2          Cafe, Mexican, Italian         800  Buffet  
3      South Indian, North Indian         300  Buffet  
4        North Indian, Rajasthani         600  Buffet

In [10]:
# Delete null values of Name column as name is the primary identifier of the dataset

hotels.dropna(subset=['Name'],axis=0,inplace=True)

# Replace null values of online order with NA

hotels['Online_Order'].fillna('NA',inplace=True)

# Replace null values of book_table with NA

hotels['Book_Table'].fillna('NA',inplace=True)

# Replace null values of location to NA

hotels['Location'].fillna('NA',inplace=True)

# Replace null values of rest_type to NA

hotels['Rest_Type'].fillna('NA',inplace=True)

# Replace null values of dishliked to NA

hotels['Dish_Liked'].fillna('NA',inplace=True)

# Replace null values of cuisines to NA

hotels['Cuisines'].fillna('NA',inplace=True)

# Replace null values of type to NA

hotels['Type'].fillna('NA',inplace=True)

# Replace null values of rating to zero as it is a numerical datatype

hotels['Rating'].fillna(0,inplace=True)

In [11]:
# Convert "Votes" and "Approx_Cost" columns to numeric with NaN for non-numeric values

hotels['Votes'] = pd.to_numeric(hotels['Votes'], errors='coerce')
hotels['Approx_Cost'] = pd.to_numeric(hotels['Approx_Cost'], errors='coerce')

In [12]:
# Remove rows with missing values in the "Votes" and "Approx_Cost" columns

hotels = hotels.dropna(subset=['Votes', 'Approx_Cost'])

In [13]:
hotels.isnull().sum()

Name            0
Online_Order    0
Book_Table      0
Rating          0
Votes           0
Location        0
Rest_Type       0
Dish_Liked      0
Cuisines        0
Approx_Cost     0
Type            0
dtype: int64

In [14]:
# Convert the columns to integers

hotels['Votes'] = hotels['Votes'].astype(np.int64)
hotels['Approx_Cost'] = hotels['Approx_Cost'].astype(np.int64)

In [15]:
hotels.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44454 entries, 0 to 56248
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Name          44454 non-null  object
 1   Online_Order  44454 non-null  object
 2   Book_Table    44454 non-null  object
 3   Rating        44454 non-null  object
 4   Votes         44454 non-null  int64 
 5   Location      44454 non-null  object
 6   Rest_Type     44454 non-null  object
 7   Dish_Liked    44454 non-null  object
 8   Cuisines      44454 non-null  object
 9   Approx_Cost   44454 non-null  int64 
 10  Type          44454 non-null  object
dtypes: int64(2), object(9)
memory usage: 4.1+ MB


In [16]:
# Identifying Duplicate Data in a Dataset

hotels.duplicated().sum()

15420

In [17]:
# Drop all the duplicate values

hotels.drop_duplicates(inplace=True)
hotels.duplicated().sum()

0

In [18]:
hotels['Name'].unique()

array(['Jalsa', 'Spice Elephant', 'San Churro Cafe', ...,
       'Calcutta North Indian Meals', 'Nawabs Empire', 'SeeYa Restaurant'],
      dtype=object)

In [ ]:
# Removing irrelevant text from columns

hotels['Name'] = hotels['Name'].str.replace('RATED|Rated', '')
hotels['Online_Order'] = hotels['Online_Order'].str.replace('RATED|Rated', '')
hotels['Book_Table'] = hotels['Book_Table'].str.replace('RATED|Rated', '')
hotels['Location'] = hotels['Location'].str.replace('RATED|Rated', '')
hotels['Rest_Type'] = hotels['Rest_Type'].str.replace('RATED|Rated', '')
hotels['Dish_Liked'] = hotels['Dish_Liked'].str.replace('RATED|Rated', '')
hotels['Cuisines'] = hotels['Cuisines'].str.replace('RATED|Rated', '')
hotels['Type'] = hotels['Type'].str.replace('RATED|Rated', '')

In [ ]:
hotels['Online_Order'].unique()

In [ ]:
hotels['Rating'].unique()

In [ ]:
# Handle the irrelevant values

hotels['Online_Order'] = hotels['Online_Order'].apply(lambda x: 'No' if x not in ['Yes', 'No'] else x)
hotels['Rating'] = hotels['Rating'].astype(str)
hotels = hotels[~hotels['Rating'].str.contains('[a-zS]')]
hotels['Rating'] = hotels['Rating'].replace(['-', 'NEW'], '0')
hotels['Rating'] = hotels['Rating'].str.replace('/5', '')

In [ ]:
hotels['Rating'].unique()

In [ ]:
# Remove the unknown character from the dataset and export it to "zomatocleaned.csv"

hotels['Name']=hotels['Name'].str.replace('[Ãx][^A-Za-z]+','',regex=True)

In [ ]:
# Export cleaned Dataset to new csv file named "zomatocleaned.csv"

hotels.to_csv('zomato_cleaned.csv', index=True)